src/evaluate.py

```diff
@@ -0,0 +1,184 @@
+            return 100  # Return 100 if successfully parsed and not None
+        return 0  # This case handles empty or equivalent to None YAML content
+    except yaml.YAMLError:
+        # Return 0 if there is any error in parsing the YAML
+        return 0
+
+
+def levenshtein_distance(expected_answer, model_output):
+    """
+    Calculate the Levenshtein distance between two YAML configurations.
+
+    Parameters:
+        expected_answer (str or dict): The correct YAML configuration.
+        model_output (str or dict): The YAML configuration generated by the model.
+
+    Returns:
+        int: The Levenshtein distance between the two YAML configurations.
+    """
+    # Convert dictionaries to YAML strings if they are not already strings
+    if isinstance(expected_answer, dict):
+        expected_answer = yaml.dump(expected_answer)
+    if isinstance(model_output, dict):
+        model_output = yaml.dump(model_output)
+
+    # Calculate and return the Levenshtein distance
+    return Levenshtein.distance(expected_answer, model_output)
+
+
+def evaluate_model(evaluation_tokenizer, trained_model, processed_data, metric_types):
+    question_list = []
+    expected_answer_list = []
+    generated_answer_list = []
+    data_len = len(processed_data)
+    for data in processed_data:
+        question = data["question"]
+        question_list.append(question)
+        expected_answer = data["answer"]
+        expected_answer_list.append(expected_answer)
+        answer = generate_answer(question, evaluation_tokenizer, trained_model)
+        generated_answer_list.append(answer)
+
+    results = {}
+    for metric_type in metric_types:
+        if metric_type == "EM":
+            results["EM"] = (
+                sum(
+                    exact_match_score(et, pt)
+                    for et, pt in zip(expected_answer_list, generated_answer_list)
+                )
+                / data_len
+            )
+        elif metric_type == "Schema Correct":
+            results["Schema_Correct"] = (
+                sum(schema_correct(pt) for pt in generated_answer_list) / data_len
+            )
+        elif metric_type == "Levenshtein Distance":
+            results["Levenshtein_Distance"] = (
+                sum(
+                    levenshtein_distance(et, pt)
+                    for et, pt in zip(expected_answer_list, generated_answer_list)
+                )
+                / data_len
+            )
+
+    return results
+
+
+def generate_answer(question, evaluation_tokenizer, trained_model):
+    # Tokenize the question
+    inputs = evaluation_tokenizer.encode(question, return_tensors="pt")
+
+    # Generate the answer
+    outputs = trained_model.generate(inputs, max_length=500)
+    answer = evaluation_tokenizer.decode(outputs[0], skip_special_tokens=True)
+
+    # Removing the repeated question if it appears in the answer
+    cleaned_answer = answer.replace(question, "").strip()
+
+    return cleaned_answer
```

Comment:
IMHO it might be better to separate creating the data set to evaluate and to calculate the actual metrics. Then the metrics can be re-done without having to re-compute the answers from the model.

Reply:
Sorry, I did not fully get it, the `generate_answer` function is only called once for each question in the testing data.  Do you mean that we should evaluate/validate by generating the evaluation metrics during the training to prevent overfitting ?

Reply:
I mean that the generated answers should be cached in a file and the evaluation of the model should be done on the cached files so the metrics can be adjusted without asking the model again which would take longer.